### PR TITLE
Handle 'best_available' QoS policies

### PR DESCRIPTION
**Pull request by jacobperron**
Connects to https://github.com/ros2/rmw/pull/320

If users set certain QoS policies to BEST_AVAILABLE, then we query the graph for endpoint info.
The policy is updated such that it is compatible with all endpoints while maintaining the
highest possible level of service.

### CODE REVIEW

File: rmw_fastrtps_cpp/src/rmw_publisher.cpp [hunk 1]
```diff
@@ -72,6 +75,16 @@ rmw_create_publisher(
     node->implementation_identifier,
     eprosima_fastrtps_identifier,
     return nullptr);
+  RMW_CHECK_ARGUMENT_FOR_NULL(qos_policies, nullptr);
```
Comment (j-rivero) [inline rmw_fastrtps_cpp/src/rmw_publisher.cpp:78]: If I understand correctly, this is going to exit the function in `qos_policies` is a nullptr. Is this not a case supported by the API? I can not find documentation for it in https://docs.ros2.org/foxy/api/rmw/rmw_8h.html#a149b0600fa9ce696c83a63f0e1e806ce

**Comment by j-rivero** [comment]
I see a couple of functions `rmw_create_client` and `rmw_create_service` that also deal with `qos_profiles`.  Do we need to implement the logic for them too?

File: rmw_fastrtps_cpp/src/rmw_publisher.cpp (hunk 1 above)
Comment (wjwwood) [reply rmw_fastrtps_cpp/src/rmw_publisher.cpp:78]: Make sure to look at the latest version: https://docs.ros2.org/galactic/api/rmw/rmw_8h.html#ac74022d90eb22d95b082d4808670f8a9

It has better docs 👍

Comment (j-rivero) [reply rmw_fastrtps_cpp/src/rmw_publisher.cpp:78]: aaaaaah, thanks! Tried rolling and fallback to foxy but completely forgot about Galactic.

**Comment by jacobperron** [comment]
Good point about services and clients. I'm not sure if we should support matching for them or not. At the very least we should define what happens if we provide 'best effort' policies.

[We have explicit policies](https://github.com/ros2/rmw/blob/01d0db60c5bac242fe6f8a6d49494512f1e20c61/rmw/include/rmw/qos_profiles.h#L64-L75) for service and clients, and I think it is rare, and perhaps not recommended, to create services and clients that are not using these policies. So, I'm not sure if we should try handling 'best effort' policies for services. I need to think about it a little more.

**Review by ivanpauno — APPROVED** [review]
LGTM!

About what to do with services, I'm not sure.
Maybe trying to do the same we do with publishers, using a default or failing sound all okay to me.

**Comment by j-rivero** [comment]
A final though about design: since we are adding almost the same code in every `rmw_` implementation, something like:

```c
  rmw_qos_profile_t adapted_qos_policies = *qos_policies;
  rmw_ret_t ret = rmw_dds_common::qos_profile_get_best_available_for_topic_publisher(
    node, topic_name, &adapted_qos_policies, rmw_get_subscriptions_info_by_topic);
  if (RMW_RET_OK != ret) {
    return nullptr;
  }
```

Would it make sense to move it to a different layer of the stack (`rmw_common?` or `rmw`) or does it fit better into this one of rmw implementations?

**Review by wjwwood — APPROVED** [review]
lgtm

I think the boilerplate that is there is not likely to change, it's essentially just calling the common function. So I wouldn't try to de-duplicate it I think, because if we did it would probably involve a macro and I don't think that's a a good idea personally. But maybe a c++ function that returns the adapted qos policy might shorten it slightly.

**Comment by jacobperron** [comment]
I'm not sure how much more of the boilerplate we could move into a common place (e.g. rmw_dds_common). Like William points out, we could try to change the signature of `rmw_dds_common::qos_profile_get_best_available_for_topic_publisher` to slim it down a bit, maybe using a `std::optional` or `std::tuple` to communicate an error? I don't think it would change the call site much though, e.g.

```c++
  std::optional<rmw_qos_profile_t adapted_qos_policies =
    rmw_dds_common::qos_profile_get_best_available_for_topic_publisher(
      node, topic_name, qos_policies, rmw_get_subscriptions_info_by_topic);
  if (!adapted_qos_policies) {
    return nullptr;
  }
```

**Comment by jacobperron** [comment]
I'm proposing we "ignore" the best available policy for services by defaulting to values in `rmw_qos_profile_services_default`. [9bb3bdf](https://github.com/ros2/rmw_fastrtps/pull/598/commits/9bb3bdf960d34ddff5061d55338554990e0bca65)

**Review by wjwwood — APPROVED** [review]

**Review by j-rivero — APPROVED** [review]

**Comment by j-rivero** [comment]
I see some failures in my local workspace related to fastrtps_dynamic. Not sure if they are related to my local setup:
```
space in code/ros2/ws_qos_adaptative ❯ colcon test-result
build/test_quality_of_service/Testing/20220426-1502/Test.xml: 22 tests, 0 errors, 3 failures, 0 skipped
build/test_quality_of_service/test_results/test_quality_of_service/test_deadline__rmw_fastrtps_dynamic_cpp.gtest.xml: 1 test, 1 error, 0 failures, 0 skipped
build/test_quality_of_service/test_results/test_quality_of_service/test_lifespan__rmw_fastrtps_dynamic_cpp.gtest.xml: 1 test, 1 error, 0 failures, 0 skipped
build/test_quality_of_service/test_results/test_quality_of_service/test_liveliness__rmw_fastrtps_dynamic_cpp.gtest.xml: 1 test, 1 error, 0 failures, 0 skipped

...

25: test_subscriber_terminates_in_a_finite_amount_of_time[BoundedPlainSequences] (test_communication.TestPublisherSubscriber)
25: Test that the subscriber terminates after a finite amount of time. ... [INFO] [test_publisher-8]: process started with pid [757421]
25: [INFO] [test_subscriber-9]: process started with pid [757423]
13: [INFO] [1650985483.968604443] [publisher]: publisher: publishing 'publisher: testing 22'                    
13: [INFO] [1650985483.968782138] [subscriber]: subscriber: subscriber heard [publisher: testing 22]
13: [INFO] [1650985484.218879834] [publisher]: publisher: publishing 'publisher: testing 23'                    
13: [INFO] [1650985484.219087234] [subscriber]: subscriber: subscriber heard [publisher: testing 23]
25: [INFO] [test_subscriber-9]: process has finished cleanly [pid 757423]
25: [INFO] [test_publisher-8]: sending signal 'SIGINT' to process[test_publisher-8]
25: [test_publisher-8] [INFO] [1650985484.277540590] [rclcpp]: signal_handler(signum=2)
13: [INFO] [1650985484.468542173] [publisher]: publisher: publishing 'publisher: testing 24'                    
13: [INFO] [1650985484.468859520] [subscriber]: subscriber: subscriber heard [publisher: testing 24]
13: -- run_test.py: return code -11                                                                             
13: -- run_test.py: generate result file '/home/jrivero/code/ros2/ws_qos_adaptative/build/test_quality_of_service/test_results/test_quality_of_service/test_deadline__rmw_fastrtps_dynamic_cpp.gtest.xml' with failed test
13: -- run_test.py: verify result file '/home/jrivero/code/ros2/ws_qos_adaptative/build/test_quality_of_service/test_results/test_quality_of_service/test_deadline__rmw_fastrtps_dynamic_cpp.gtest.xml'
13/22 Test #13: test_deadline__rmw_fastrtps_dynamic_cpp .........***Failed   10.14 sec
...
26: Starting test run test_communication.test_requester_replier__rclcpp__rmw_connextdds_Release.launch_tests[Arrays]
26: [INFO] [launch]: All log files can be found below /home/jrivero/.ros/log/2022-04-26-16-21-08-673136-space-727240
26: [INFO] [launch]: Default logging verbosity is set to INFO
26: [INFO] [daemon-stop-1]: process started with pid [727243]
26: [INFO] [daemon-stop-1]: process has finished cleanly [pid 727243]                                           
26: test_requester_finishes_in_a_finite_amount_of_time[Arrays] (test_communication.TestRequesterReplier)
26: Test that the second executable terminates after a finite amount of time. ... [INFO] [test_replier-2]: process started with pid [727246]
26: [INFO] [test_requester-3]: process started with pid [727248]
26: [INFO] [test_requester-3]: process has finished cleanly [pid 727248]                                        
26: [INFO] [test_replier-2]: sending signal 'SIGINT' to process[test_replier-2]
26: [test_replier-2] [INFO] [1650982869.257146156] [rclcpp]: signal_handler(signum=2)
15: -- run_test.py: return code -11                                                                             
15: -- run_test.py: generate result file '/home/jrivero/code/ros2/ws_qos_adaptative/build/test_quality_of_service/test_results/test_quality_of_service/test_liveliness__rmw_fastrtps_dynamic_cpp.gtest.xml' with failed test
15: -- run_test.py: verify result file '/home/jrivero/code/ros2/ws_qos_adaptative/build/test_quality_of_service/test_results/test_quality_of_service/test_liveliness__rmw_fastrtps_dynamic_cpp.gtest.xml'
15/22 Test #15: test_liveliness__rmw_fastrtps_dynamic_cpp .......***Failed    8.14 sec
...
5: OK
25: test_subscriber_terminates_in_a_finite_amount_of_time[UnboundedSequences] (test_communication.TestPublisherSubscriber)
25: Test that the subscriber terminates after a finite amount of time. ... [INFO] [test_publisher-35]: process started with pid [757704]
25: [INFO] [test_subscriber-36]: process started with pid [757706]
14: [INFO] [1650985494.360158289] [publisher]: publisher: publishing 'publisher: testing 78'                 
14: [INFO] [1650985494.360315324] [subscriber]: subscriber: subscriber heard [publisher: testing 78]
14: [INFO] [1650985494.485499664] [publisher]: publisher: publishing 'publisher: testing 79'                 
14: [INFO] [1650985494.485702816] [subscriber]: subscriber: subscriber heard [publisher: testing 79]
25: [INFO] [test_subscriber-36]: process has finished cleanly [pid 757706]                                   
25: [INFO] [test_publisher-35]: sending signal 'SIGINT' to process[test_publisher-35]
25: [test_publisher-35] [INFO] [1650985494.594500480] [rclcpp]: signal_handler(signum=2)
14: [INFO] [1650985494.610140462] [publisher]: publisher: publishing 'publisher: testing 80'
14: -- run_test.py: return code -11                                                                          
14: -- run_test.py: generate result file '/home/jrivero/code/ros2/ws_qos_adaptative/build/test_quality_of_service/test_results/test_quality_of_service/test_lifespan__rmw_fastrtps_dynamic_cpp.gtest.xml' with failed test
14: -- run_test.py: verify result file '/home/jrivero/code/ros2/ws_qos_adaptative/build/test_quality_of_service/test_results/test_quality_of_service/test_lifespan__rmw_fastrtps_dynamic_cpp.gtest.xml'
14/22 Test #14: test_lifespan__rmw_fastrtps_dynamic_cpp .........***Failed   10.14 sec
```

We  can wait for Jenkins valid results before investing time in inspecting the errors.

**Comment by jacobperron** [comment]
CI is green: https://github.com/ros2/rmw/pull/320#issue-1205954246

**Comment by ivanpauno** [comment]
> I see some failures in my local workspace related to fastrtps_dynamic. Not sure if they are related to my local setup:

See https://github.com/ros2/rmw_fastrtps/issues/599

**Pull request merged**